src/Behat/Testwork/Argument/MixedArgumentOrganiser.php

```diff
@@ -174,27 +174,78 @@ private function prepareNamedArguments(array $parameters, array $namedArguments)
     }
 
     /**
-     * Captures argument value based on their respective typehints.
+     * Captures argument values for typehinted arguments based on the given candidates.
      *
-     * Note ; as it is keeping in mind that $typeHintedArgument already has the
-     * right keys matching the correct parameter, it just iterates over this
-     * array in order to mark each arguments as "defined".
+     * This method attempts to match up the best fitting arguments to each constructor argument.
      *
-     * @see https://github.com/Behat/Behat/pull/993#issuecomment-275669510
+     * This case specifically fixes the issue where a constructor asks for a parent and child class,
+     * as separate arguments, but both arguments could satisfy the first argument,
+     * so they would both be passed in (overwriting each other).
      *
-     * @param mixed[] $typehintedArguments
+     * This will ensure that the children (exact class matches) are mapped first, and then other dependencies
+     * are mapped sequentially (to arguments which they are an `instanceof`).
      *
-     * @return mixed[]
+     * As such, this requires two passes of the $parameters array to ensure it is mapped as accurately as possible.
+     *
+     * @param ReflectionParameter[] $parameters          Reflection Parameters (constructor argument requirements)
+     * @param mixed[]               $typehintedArguments Resolved arguments
+     *
+     * @return mixed[] Ordered list of arguments, index is the constructor argument position, value is what will be injected
      */
-    private function prepareTypehintedArguments(array $typehintedArguments)
+    private function prepareTypehintedArguments(array $parameters, array $typehintedArguments)
     {
-        foreach (array_keys($typehintedArguments) as $num) {
+        $arguments = [];
+
+        $candidates = $typehintedArguments;
+
+        foreach ($parameters as $num => $parameter) {
+            if ($this->isArgumentDefined($num)) {
+                continue;
+            }
+
+            foreach ($candidates as $candidateIndex => $candidate) {
+                $reflectionClass = $parameter->getClass();
+
+                if (!$reflectionClass || !$reflectionClass->isInstance($candidate)) {
+                    continue;
+                }
+
+                if ($reflectionClass->getName() === get_class($candidate)) {
+                    $arguments[$num] = $candidate;
+                    
             $this->markArgumentDefined($num);
+
+                    unset($candidates[$candidateIndex]);
+
+                    break 1;
         }
+            }
+        }
+
+        foreach ($parameters as $num => $parameter) {
```

Comment:
This foreach block looks duplicated to me, the only difference is the `if ($reflectionClass->getName() === get_class($candidate))` statement. It would be nice if you could eliminate this repetition.